src/scene/shader-lib/utils.js

```diff
@@ -31,19 +31,20 @@ function createShader(device, vsName, fsName, useTransformFeedback = false) {
  * @param {string} fsCode - The fragment shader code.
  * @param {string} uniqueName - Unique name for the shader.
  * @param {boolean} [useTransformFeedback] - Whether to use transform feedback. Defaults to false.
- * @param {string} [fragmentPreamble] - An optional 'preamble' string for the fragment shader. Defaults
- * to ''.
+ * @param {Object<string, string>} [attributes] - Object detailing the mapping of vertex shader
+ * attribute names to semantics SEMANTIC_*. This enables the engine to match vertex buffer data as
+ * inputs to the shader. Defaults to undefined, which generates the default attributes.
  * @returns {Shader} The newly created shader.
  */
-function createShaderFromCode(device, vsCode, fsCode, uniqueName, useTransformFeedback = false, fragmentPreamble = '') {
+function createShaderFromCode(device, vsCode, fsCode, uniqueName, useTransformFeedback = false, attributes) {
```

Comment:
Might be worth swapping the last two arguments around since attributes is (presumably) much more likely to get used compared to useTransformFeedback?

Reply:
I thought about it, but users are likely using the transform feedback parameter.
Maybe I can check types and warn the users.

Reply:
OIC, thought this was internal. Probably best to leave it then.

Reply:
ok done, and added assert to test for incorrect calls to the function